userspace/engine/rule_loader.cpp

```diff
@@ -326,19 +273,89 @@ static void resolve_macros(
 	}
 }
 
-// note: there is no visibility order between filter conditions and lists
-static shared_ptr<ast::expr> parse_condition(
-	string condition,
-	indexed_vector<rule_loader::list_info>& lists)
+struct macro_list_inserter
 {
-	for (auto &l : lists)
+	regex_t re{};
+	macro_list_inserter()
+	{
+		auto rgx = std::string() + "[\\(][_a-z0-9]+[\\)]";
+		if(regcomp(&re, rgx.c_str(), REG_EXTENDED) != 0){ ASSERT(false);}
+	}
+
+	~macro_list_inserter(){regfree(&re);}
+	std::string insert_lists(const std::string &cond,
+			  indexed_vector<rule_loader::list_info>& lists) const
 	{
-		if (resolve_list(condition, l))
+		static delim_chars blanks = {" \t\n\r"};
+		static delim_chars delims = {std::string(" \t\n\r") + "(),="};
+
+		std::string ret;
+		regmatch_t re_match;
+		size_t start = 0;
+
+		while(start < cond.size() && regexec(&re, cond.c_str() + start, 1, &re_match, 0) == 0)
 		{
-			l.used = true;
+			if((re_match.rm_so == 0 || delims.contains(cond[start + re_match.rm_so -1]))
+			   && ( start + re_match.rm_eo >= cond.size() || delims.contains(cond[start + re_match.rm_eo]) ))
+			{
+				auto *li  = lists.at(cond.substr(start + re_match.rm_so+1, re_match.rm_eo - re_match.rm_so-2));
+				if (li != nullptr)
+				{
+					li->used = true;
+					ret += cond.substr(start,  re_match.rm_so);
+					ret += "(";
```

Comment:
I didn't run this, but I think you may be skipping cases when lists are referenced in other lists like:
```
evt.type in (open, some_types_list, connect)
```
Unfortunately, this is legit and supported.

Reply:
fixed